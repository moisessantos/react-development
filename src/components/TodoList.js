import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Item from './Item';

class TodoList extends Component {
  render() {
    return (
      <div>
        <h1>This is my todo list</h1>
        { this.props.items.map((item, index) => <Item key={index}>{item}</Item>) }
      </div>
    );
  }
}

TodoList.propTypes = {
  items: PropTypes.arrayOf(PropTypes.string)
};

export default TodoList;
