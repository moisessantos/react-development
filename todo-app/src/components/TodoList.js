import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Item from './Item';

class TodoList extends Component {
  constructor(props) {
    super(props)
    this.state = {
      currentItems: props.items,
    };

    this.addTodo = this.addTodo.bind(this);
    this.removeTodo = this.removeTodo.bind(this);

    this.ref = React.createRef();
  }
  render() {
    return (
      <div className="container">
        <h1>This is my todo list</h1>
        {this.renderAddTodo()}     
        { this.state.currentItems
            .map((item, index) => <Item key={index} onDelete={this.removeTodo}>{item}</Item>) }
      </div>
    );
  }

  renderAddTodo() {
    return (
      <div className="addTodo">
        <input type="text" ref={this.ref} />
        <button onClick={ this.addTodo }>Add todo</button>
      </div> 
    );
  }

  addTodo() {
    if(this.state.currentItems.includes(this.ref.current.value)){
      return;
    }
    
    this.state.currentItems.push(this.ref.current.value);
    this.ref.current.value = '';
    
    this.setState({
      currentItems: this.state.currentItems,
    });
  }

  removeTodo(todo) {
    this.setState({
      currentItems: this.state.currentItems.filter(x => x !== todo),
    });
  }
}

TodoList.propTypes = {
  items: PropTypes.arrayOf(PropTypes.string)
};

export default TodoList;
