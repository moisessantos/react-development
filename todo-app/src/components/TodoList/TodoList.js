import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Item } from '..';

class TodoList extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      currentItems: props.items,
    };
    this.addTodo = this.addTodo.bind(this);
  }
  render() {
    return (
      <div className="container">
        <h1>This is my todo list</h1>
        {this.renderAddTodo()}     
        { this.state.currentItems
            .map((item, index) => <Item key={index} onDelete={this.removeTodo.bind(this)}>{item}</Item>) }
      </div>
    );
  }

  renderAddTodo() {
    return (
      <div className="addTodo">
        <input type="text" />
        <button onClick={ (e) => this.addTodo(e.target.previousSibling) }>Add todo</button>
      </div> 
    );
  }

  addTodo(element) {
    if(this.state.currentItems.includes(element.innerHTML)){
      return;
    }
    
    this.state.currentItems.push(element.innerHTML);
    element.innerHTML = '';
    
    this.setState({
      currentItems: this.state.currentItems,
    });
  }

  removeTodo(todo) {
    this.setState({
      currentItems: this.state.currentItems.filter(x => x !== todo),
    });
  }
}

TodoList.propTypes = {
  items: PropTypes.arrayOf(PropTypes.string)
};

export default TodoList;
