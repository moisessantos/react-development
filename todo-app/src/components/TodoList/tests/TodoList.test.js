import React from 'react';
import { shallow } from 'enzyme';
import TodoList from '../';

describe('TodoList component', () => {
  it('should render snapshot', () => {
  const wrapper = shallow(<TodoList items = { ['Get up', 'Do homework'] } />);
    expect(wrapper).toMatchSnapshot();
  });
  
  it('should call addTodo and add item when is not in the list', () => {
    const todo = 'New todo'; 
    const mock = jest.fn();
    const wrapper = shallow(<TodoList items = { ['Get up', 'Do homework'] } />);

    wrapper.find('input').text(todo);
    wrapper.find('button').simulate('click',{ target: { previousSibling: 
      { 
        get innerHTML(){ return todo}, 
        set innerHTML(value) { mock(value)}}
      }
    });
    expect(wrapper.state('currentItems')[2]).toBe(todo);
    expect(mock).toBeCalledTimes(1);
    expect(mock).toBeCalledWith('');
  });

  it('should call addTodo and not add item when is in the list', () => {
    const todo = 'Get up'; 
    const mock = jest.fn();
    const wrapper = shallow(<TodoList items = { ['Get up', 'Do homework'] } />);

    wrapper.find('input').text(todo);
    wrapper.find('button').simulate('click',{ target: { previousSibling: 
      { 
        get innerHTML(){ return todo}, 
        set innerHTML(value) { mock(value)}}
      }
    });
    expect(wrapper.state('currentItems').length).toBe(2);
    expect(mock).toBeCalledTimes(0);
  });
});
