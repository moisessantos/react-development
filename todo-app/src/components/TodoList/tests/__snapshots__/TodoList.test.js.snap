// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component should render snapshot 1`] = `
<div
  className="container"
>
  <h1>
    This is my todo list
  </h1>
  <div
    className="addTodo"
  >
    <input
      type="text"
    />
    <button
      onClick={[Function]}
    >
      Add todo
    </button>
  </div>
  <Memo(Item)
    key="0"
    onDelete={[Function]}
  >
    Get up
  </Memo(Item)>
  <Memo(Item)
    key="1"
    onDelete={[Function]}
  >
    Do homework
  </Memo(Item)>
</div>
`;
